### PR TITLE
Adds feature to filter monkey by intent category

### PR DESCRIPTION
**Pull request by jackSzm**
This PR Adds feature to filter monkey by intent category. 

It's done by adding `'-c' INTENT_FILTER` to the shell command. 
Default to the monkeys normal categories if `-c` not specified

Paired with @blundell

### CODE REVIEW

File: gradle-android-command-plugin/src/main/groovy/com/novoda/gradle/command/Monkey.groovy [hunk 1]
```diff
@@ -15,12 +16,29 @@ class Monkey extends AdbTask {
 
     @TaskAction
     void exec() {
-        def arguments = ['shell', 'monkey', '-p', packageName, '-v', getEvents()]
+        def arguments = ['shell', 'monkey', '-p', packageName]
```
Comment (jonreeve) [inline gradle-android-command-plugin/src/main/groovy/com/novoda/gradle/command/Monkey.groovy]: I might just split the '-p' and packageName ones out onto the next line too, for consistency

File: gradle-android-command-plugin/src/main/groovy/com/novoda/gradle/command/Monkey.groovy [hunk 2]
```diff
@@ -15,12 +16,29 @@ class Monkey extends AdbTask {
...
     }
 
+    def getFormattedCategories() {
+        def formattedCategories = []
+        def categories = getCategories()
+        categories.each {
+            formattedCategories.add("-c " + it)
+        }
+        formattedCategories
+    }
```
Comment (jonreeve) [inline gradle-android-command-plugin/src/main/groovy/com/novoda/gradle/command/Monkey.groovy]: Seems very Java-ish. Can't you just transform getCategories() with the collect function and a closure? Think this would be a simple single line function...

File: gradle-android-command-plugin/src/main/groovy/com/novoda/gradle/command/Monkey.groovy [hunk 3]
```diff
@@ -15,12 +16,29 @@ class Monkey extends AdbTask {
...
+
+    private getCategories() {
+        if (categories instanceof Closure)
+            categories = categories.call()
```
Comment (jonreeve) [inline gradle-android-command-plugin/src/main/groovy/com/novoda/gradle/command/Monkey.groovy]: This code seems very... braceless. Not just your addition mind, the whole thing. Scary.

Comment (blundell) [inline gradle-android-command-plugin/src/main/groovy/com/novoda/gradle/command/Monkey.groovy]: true, was copying the methods belows evilness :arrow_down:

File: sample/src/main/java/com/novoda/gradle/command/MonkeyActivity.java [hunk 4]
```diff
@@ -0,0 +1,25 @@
+package com.novoda.gradle.command;
+
+import android.app.Activity;
+import android.os.Bundle;
+import android.view.Menu;
+import android.view.MenuItem;
+
+public class MonkeyActivity extends Activity {
```
Comment (jonreeve) [inline sample/src/main/java/com/novoda/gradle/command/MonkeyActivity.java]: :+1:

File: gradle-android-command-plugin/src/main/groovy/com/novoda/gradle/command/Monkey.groovy (hunk 2 above)
Comment (blundell) [inline gradle-android-command-plugin/src/main/groovy/com/novoda/gradle/command/Monkey.groovy]: now you have said the word transform it makes sense!

```groovy
def getFormattedCategories() {
    categories.collect { "-c " + it }
}

Comment (jonreeve) [inline gradle-android-command-plugin/src/main/groovy/com/novoda/gradle/command/Monkey.groovy]: yup, exactly :smile: Closures ftw

File: gradle-android-command-plugin/src/main/groovy/com/novoda/gradle/command/AndroidCommandPluginExtension.groovy [hunk 5]
```diff
@@ -77,6 +78,10 @@ public class AndroidCommandPluginExtension {
         System.properties['seed'] ?: seed
     }
 
+    def getCategories() {
+        System.properties['categories'] ?: categories ?: ''
```
Comment (ouchadam) [inline gradle-android-command-plugin/src/main/groovy/com/novoda/gradle/command/AndroidCommandPluginExtension.groovy]: double ternary :fearful: is empty ok?

File: gradle-android-command-plugin/src/main/groovy/com/novoda/gradle/command/Monkey.groovy (hunk 1 above)
Comment (ouchadam) [inline gradle-android-command-plugin/src/main/groovy/com/novoda/gradle/command/Monkey.groovy]: :+1: for separating the flags with their inputs

File: sample/src/main/java/com/novoda/gradle/command/MainActivity.java [hunk 6]
```diff
@@ -27,15 +25,4 @@ public void onClick(final View v) {
...
-
-    @Override
-    public boolean onCreateOptionsMenu(Menu menu) {
-        getMenuInflater().inflate(R.menu.main, menu);
-        return true;
-    }
-
-    @Override
-    public boolean onOptionsItemSelected(MenuItem item) {
-        throw new RuntimeException("forced error to test monkey failure.");
```
Comment (ouchadam) [inline sample/src/main/java/com/novoda/gradle/command/MainActivity.java]: :+1:

**Comment by ouchadam** [comment]
few small comments, agree with @jonreeve otherwise LGTM :+1:

File: gradle-android-command-plugin/src/main/groovy/com/novoda/gradle/command/Monkey.groovy (hunk 1 above)
Comment (jackSzm) [inline gradle-android-command-plugin/src/main/groovy/com/novoda/gradle/command/Monkey.groovy]: :+1:

File: gradle-android-command-plugin/src/main/groovy/com/novoda/gradle/command/Monkey.groovy (hunk 2 above)
Comment (jackSzm) [inline gradle-android-command-plugin/src/main/groovy/com/novoda/gradle/command/Monkey.groovy]: That's awesome ! Done :+1:

File: gradle-android-command-plugin/src/main/groovy/com/novoda/gradle/command/AndroidCommandPluginExtension.groovy (hunk 5 above)
Comment (jackSzm) [inline gradle-android-command-plugin/src/main/groovy/com/novoda/gradle/command/AndroidCommandPluginExtension.groovy]: yeah that's what we want. Empty in Monkey world means default :D

File: gradle-android-command-plugin/src/main/groovy/com/novoda/gradle/command/Monkey.groovy (hunk 3 above)
Comment (jackSzm) [inline gradle-android-command-plugin/src/main/groovy/com/novoda/gradle/command/Monkey.groovy]: Good point ! Done :+1:

File: gradle-android-command-plugin/src/main/groovy/com/novoda/gradle/command/AndroidCommandPluginExtension.groovy (hunk 5 above)
Comment (ouchadam) [inline gradle-android-command-plugin/src/main/groovy/com/novoda/gradle/command/AndroidCommandPluginExtension.groovy]: maybe a constant for `DEFAULT` would be help :smile:

File: gradle-android-command-plugin/src/main/groovy/com/novoda/gradle/command/Monkey.groovy [hunk 7]
```diff
@@ -6,30 +6,49 @@ class Monkey extends AdbTask {
...
         super.handleCommandOutput(text)
-        if (text.contains("Monkey aborted"))
+        if (text.contains("Monkey aborted")) {
             throw new GroovyRuntimeException('Monkey run failed')
+        }
     }
 
     @TaskAction
     void exec() {
-        def arguments = ['shell', 'monkey', '-p', packageName, '-v', getEvents()]
```
Comment (ouchadam) [inline gradle-android-command-plugin/src/main/groovy/com/novoda/gradle/command/Monkey.groovy]: :+1:

File: gradle-android-command-plugin/src/main/groovy/com/novoda/gradle/command/AndroidCommandPluginExtension.groovy (hunk 5 above)
Comment (jonreeve) [inline gradle-android-command-plugin/src/main/groovy/com/novoda/gradle/command/AndroidCommandPluginExtension.groovy]: lol not ternary tho. Elvis operator. Still a bit strange but not nearly so objectionable as a nested ternary.

File: gradle-android-command-plugin/src/main/groovy/com/novoda/gradle/command/Monkey.groovy [hunk 8]
```diff
@@ -6,30 +6,49 @@ class Monkey extends AdbTask {
...
-        if (getSeed())
+        def arguments = ['shell', 'monkey']
+        arguments += ['-p', packageName]
```
Comment (jonreeve) [inline gradle-android-command-plugin/src/main/groovy/com/novoda/gradle/command/Monkey.groovy]: :+1:

File: gradle-android-command-plugin/src/main/groovy/com/novoda/gradle/command/Monkey.groovy [hunk 9]
```diff
@@ -6,30 +6,49 @@ class Monkey extends AdbTask {
...
             events = events.call()
+        }
         events ?: pluginEx.events
     }
 
     private getSeed() {
-        if (seed instanceof Closure)
+        if (seed instanceof Closure) {
             seed = seed.call()
+        }
```
Comment (jonreeve) [inline gradle-android-command-plugin/src/main/groovy/com/novoda/gradle/command/Monkey.groovy]: Phew. Braces :smile:

File: gradle-android-command-plugin/src/main/groovy/com/novoda/gradle/command/AndroidCommandPluginExtension.groovy (hunk 5 above)
Comment (ouchadam) [inline gradle-android-command-plugin/src/main/groovy/com/novoda/gradle/command/AndroidCommandPluginExtension.groovy]: :+1:

File: gradle-android-command-plugin/src/main/groovy/com/novoda/gradle/command/Monkey.groovy [hunk 10]
```diff
@@ -6,30 +6,49 @@ class Monkey extends AdbTask {
...
+        arguments += getFormattedCategories()
+        arguments += ['-v', getEvents()]
+        if (getSeed()) {
             arguments += ['-s', getSeed()]
+        }
         assertDeviceAndRunCommand(arguments)
     }
 
+    def getFormattedCategories() {
+        getCategories().collect { "-c " + it }
```
Comment (jonreeve) [inline gradle-android-command-plugin/src/main/groovy/com/novoda/gradle/command/Monkey.groovy]: Groovy closures, aw yiss :sunglasses:

**Comment by jonreeve** [comment]
LGTM :+1:

File: gradle-android-command-plugin/src/main/groovy/com/novoda/gradle/command/AndroidCommandPluginExtension.groovy (hunk 5 above)
Comment (ouchadam) [inline gradle-android-command-plugin/src/main/groovy/com/novoda/gradle/command/AndroidCommandPluginExtension.groovy]: wait a sec, are the `System.properties['categories']` not the same as `categories`? 
could this be `System.properties['categories'] ?: ' '`?

Comment (jackSzm) [inline gradle-android-command-plugin/src/main/groovy/com/novoda/gradle/command/AndroidCommandPluginExtension.groovy]: I believe there are three possible, scenarios here. 
Highest priority is to `System.properties[categories]`, then specified categories `categoires` in gradle file and then `default` one. 

I copied this bit from getEvents :arrow_up: and this is how I think it works

Comment (jackSzm) [inline gradle-android-command-plugin/src/main/groovy/com/novoda/gradle/command/AndroidCommandPluginExtension.groovy]: Also moved default `events` and `categories` values to be constants

**Comment by ouchadam** [comment]
:+1:

File: gradle-android-command-plugin/src/main/groovy/com/novoda/gradle/command/AndroidCommandPluginExtension.groovy [hunk 11]
```diff
@@ -5,12 +5,16 @@ import org.gradle.api.Project
 
 public class AndroidCommandPluginExtension {
 
+    static final String CATEGORIES_DEFAULT = ''
```
Comment (devisnik) [inline gradle-android-command-plugin/src/main/groovy/com/novoda/gradle/command/AndroidCommandPluginExtension.groovy]: What a default is an empty category? Would that work at all with the '-c' ? 
Should be inlined imho.

File: gradle-android-command-plugin/src/main/groovy/com/novoda/gradle/command/AndroidCommandPluginExtension.groovy [hunk 12]
```diff
@@ -61,16 +65,30 @@ public class AndroidCommandPluginExtension {
     // prefer system property over direct setting to enable commandline arguments
     @Memoized
     def getDeviceId() {
-        if (System.properties['deviceId'])
+        if (System.properties['deviceId']) {
             return System.properties['deviceId']
-        if (deviceId instanceof Closure)
+        } else if (deviceId instanceof Closure) {
```
Comment (devisnik) [inline gradle-android-command-plugin/src/main/groovy/com/novoda/gradle/command/AndroidCommandPluginExtension.groovy]: Why the `else` here? It's now inconsistent with the last return value. Needs another `else` then.

File: gradle-android-command-plugin/src/main/groovy/com/novoda/gradle/command/AndroidCommandPluginExtension.groovy [hunk 13]
```diff
@@ -61,16 +65,30 @@ public class AndroidCommandPluginExtension {
...
-    // prefer system property over direct setting to enable commandline arguments
     def getEvents() {
-        System.properties['events'] ?: events ?: 10000
+        if (System.properties['events']) {
+            return System.properties['events']
+        } else if (events) {
+            return events
+        }
+        EVENTS_DEFAULT
+    }
```
Comment (devisnik) [inline gradle-android-command-plugin/src/main/groovy/com/novoda/gradle/command/AndroidCommandPluginExtension.groovy]: What's wrong with the operator solution. If you trade 1 line for 6 lines, you better have a good reason. 
I don't see any here.

File: gradle-android-command-plugin/src/main/groovy/com/novoda/gradle/command/AndroidCommandPluginExtension.groovy [hunk 14]
```diff
@@ -61,16 +65,30 @@ public class AndroidCommandPluginExtension {
...
+
+    def getCategories() {
+        if (System.properties['categories']) {
+            return System.properties['categories']
+        } else if (categories) {
+            return categories
+        }
+        CATEGORIES_DEFAULT
```
Comment (devisnik) [inline gradle-android-command-plugin/src/main/groovy/com/novoda/gradle/command/AndroidCommandPluginExtension.groovy]: Same here.

File: gradle-android-command-plugin/src/main/groovy/com/novoda/gradle/command/Monkey.groovy (hunk 10 above)
Comment (devisnik) [inline gradle-android-command-plugin/src/main/groovy/com/novoda/gradle/command/Monkey.groovy]: Single quotes for consistency, please!

Comment (devisnik) [inline gradle-android-command-plugin/src/main/groovy/com/novoda/gradle/command/Monkey.groovy]: I'd also suggest to separate the `-c` flag from the value, similar to what is done for the other parameters.

File: gradle-android-command-plugin/src/test/groovy/com/novoda/gradle/command/AndroidCommandPluginExtensionTest.groovy [hunk 15]
```diff
@@ -25,6 +25,11 @@ class AndroidCommandPluginExtensionTest extends GroovyTestCase {
         assert extension.getSeed() == null
     }
 
+    void testDefaultCategories() {
+        def extension = createExtension()
+        assert extension.getCategories() == AndroidCommandPluginExtension.CATEGORIES_DEFAULT
```
Comment (devisnik) [inline gradle-android-command-plugin/src/test/groovy/com/novoda/gradle/command/AndroidCommandPluginExtensionTest.groovy]: I don't see any reason not to use `null` when there is no category specified.

File: gradle-android-command-plugin/src/main/groovy/com/novoda/gradle/command/AndroidCommandPluginExtension.groovy [hunk 16]
```diff
@@ -61,16 +65,30 @@ public class AndroidCommandPluginExtension {
...
+        if (System.properties['deviceId']) {
             return System.properties['deviceId']
-        if (deviceId instanceof Closure)
+        } else if (deviceId instanceof Closure) {
             return deviceId.call()
+        }
         deviceId ?: defaultDeviceId()
     }
 
-    // prefer system property over direct setting to enable commandline arguments
```
Comment (devisnik) [inline gradle-android-command-plugin/src/main/groovy/com/novoda/gradle/command/AndroidCommandPluginExtension.groovy]: I think this is still a valuable comment. Why did you remove it? Is it obvious that the ordering here is related to overriding properties on the command line?

File: gradle-android-command-plugin/src/main/groovy/com/novoda/gradle/command/Monkey.groovy [hunk 17]
```diff
@@ -6,30 +6,49 @@ class Monkey extends AdbTask {
...
+        arguments += ['-p', packageName]
+        arguments += getFormattedCategories()
+        arguments += ['-v', getEvents()]
+        if (getSeed()) {
             arguments += ['-s', getSeed()]
+        }
         assertDeviceAndRunCommand(arguments)
     }
 
+    def getFormattedCategories() {
```
Comment (devisnik) [inline gradle-android-command-plugin/src/main/groovy/com/novoda/gradle/command/Monkey.groovy]: private?

File: gradle-android-command-plugin/src/main/groovy/com/novoda/gradle/command/AndroidCommandPluginExtension.groovy (hunk 11 above)
Comment (blundell) [inline gradle-android-command-plugin/src/main/groovy/com/novoda/gradle/command/AndroidCommandPluginExtension.groovy]: Yeah from the monkey docs an empty category is treated like the default categories. Right now the default is category of launcher or monkey.

File: gradle-android-command-plugin/src/main/groovy/com/novoda/gradle/command/AndroidCommandPluginExtension.groovy (hunk 13 above)
Comment (mr-archano) [inline gradle-android-command-plugin/src/main/groovy/com/novoda/gradle/command/AndroidCommandPluginExtension.groovy]: I agree with @devisnik here: this change seems to push for a non-idiomatic Groovy solution while in reality it should have only used a constant in place of `10000`.

File: gradle-android-command-plugin/src/main/groovy/com/novoda/gradle/command/AndroidCommandPluginExtension.groovy (hunk 16 above)
Comment (mr-archano) [inline gradle-android-command-plugin/src/main/groovy/com/novoda/gradle/command/AndroidCommandPluginExtension.groovy]: Comments should be removed only if obsolete. This is not the case, right?

File: gradle-android-command-plugin/src/main/groovy/com/novoda/gradle/command/AndroidCommandPluginExtension.groovy (hunk 13 above)
Comment (jackSzm) [inline gradle-android-command-plugin/src/main/groovy/com/novoda/gradle/command/AndroidCommandPluginExtension.groovy]: Originally I haven't changed this bit, but first reviewers didn't like double elvis operator in one line :)

File: gradle-android-command-plugin/src/main/groovy/com/novoda/gradle/command/Monkey.groovy (hunk 10 above)
Comment (jackSzm) [inline gradle-android-command-plugin/src/main/groovy/com/novoda/gradle/command/Monkey.groovy]: Since every category needs to be prefixed with -c , we came up with a solution to make a separate method for this. 
I can inline this method but I am not sure if this will be clearer. 
Do you have any other suggestions ?

Comment (blundell) [inline gradle-android-command-plugin/src/main/groovy/com/novoda/gradle/command/Monkey.groovy]: Yeah @devisnik this one is slightly different in that every value requires the `-c` thats why its inside here rather than outside.

Comment (devisnik) [inline gradle-android-command-plugin/src/main/groovy/com/novoda/gradle/command/Monkey.groovy]: My suggestion was to make this return an array `['-c', category1, '-c', 'category2', ...]`. That's how the arguments are passed for the other parameters.

Comment (blundell) [inline gradle-android-command-plugin/src/main/groovy/com/novoda/gradle/command/Monkey.groovy]: can you explain some more? As we still have to iterate over `getCategories`

File: gradle-android-command-plugin/src/main/groovy/com/novoda/gradle/command/AndroidCommandPluginExtension.groovy (hunk 11 above)
Comment (devisnik) [inline gradle-android-command-plugin/src/main/groovy/com/novoda/gradle/command/AndroidCommandPluginExtension.groovy]: We shouldn't set the `-c` when there is no category.

File: gradle-android-command-plugin/src/main/groovy/com/novoda/gradle/command/AndroidCommandPluginExtension.groovy (hunk 13 above)
Comment (devisnik) [inline gradle-android-command-plugin/src/main/groovy/com/novoda/gradle/command/AndroidCommandPluginExtension.groovy]: May be you went a bit to fast here. ;)
I do not see the request to replace in the comments. 

The `?:` operator might be a bit unusual for the java native, but imho makes perfect sense here. Actually, it's made for this use case.

The change now needs to request the `events` property twice, is inconsistent regarding `if-else` and `return` and adds a lot of syntax noise. 

Wonder how we should deal with this now.

File: gradle-android-command-plugin/src/main/groovy/com/novoda/gradle/command/AndroidCommandPluginExtension.groovy (hunk 12 above)
Comment (jackSzm) [inline gradle-android-command-plugin/src/main/groovy/com/novoda/gradle/command/AndroidCommandPluginExtension.groovy]: Can't remember the reason why it has been changed. 
Will revert it then.

File: gradle-android-command-plugin/src/main/groovy/com/novoda/gradle/command/AndroidCommandPluginExtension.groovy (hunk 16 above)
Comment (jackSzm) [inline gradle-android-command-plugin/src/main/groovy/com/novoda/gradle/command/AndroidCommandPluginExtension.groovy]: When changing from double elvis operator I find it quiet obvious what the if else statement does.
If we want to go back to one-line solution then yes, I think that comment is still needed.

File: gradle-android-command-plugin/src/test/groovy/com/novoda/gradle/command/AndroidCommandPluginExtensionTest.groovy (hunk 15 above)
Comment (jackSzm) [inline gradle-android-command-plugin/src/test/groovy/com/novoda/gradle/command/AndroidCommandPluginExtensionTest.groovy]: We still use CATEGORIES_DEFAULT value when it's not specified and I think it brings more consistency across other parameters.

File: gradle-android-command-plugin/src/main/groovy/com/novoda/gradle/command/AndroidCommandPluginExtension.groovy (hunk 11 above)
Comment (blundell) [inline gradle-android-command-plugin/src/main/groovy/com/novoda/gradle/command/AndroidCommandPluginExtension.groovy]: but doing a blank -c allows the adb command to use the default categories (the same as not setting it)

File: gradle-android-command-plugin/src/main/groovy/com/novoda/gradle/command/AndroidCommandPluginExtension.groovy (hunk 13 above)
Comment (jackSzm) [inline gradle-android-command-plugin/src/main/groovy/com/novoda/gradle/command/AndroidCommandPluginExtension.groovy]: Yeah, I get the usage of `?:` operator. 

I am happy to change it back to the nested/double elvis  and bring back the comment as long as somebody would give two :+1:  and merge my changes :) 

I believe it has been already pointed out that it's a drawback of Groovy and all these other new languages that they don't have/we don't use static analysis tools and do this kind of job for us.

Comment (devisnik) [inline gradle-android-command-plugin/src/main/groovy/com/novoda/gradle/command/AndroidCommandPluginExtension.groovy]: Yes, that's true. Would make things a bit easier. 
I'm sorry that this now affects your contribution, but I have to make sure that these opinionated changes don't sneak and cause a lot of discussion when changing them later on. :smile: 

I'll give you two :+1: for reverting. :dancers:

File: gradle-android-command-plugin/src/test/groovy/com/novoda/gradle/command/AndroidCommandPluginExtensionTest.groovy (hunk 15 above)
Comment (devisnik) [inline gradle-android-command-plugin/src/test/groovy/com/novoda/gradle/command/AndroidCommandPluginExtensionTest.groovy]: I see, I still find it a bit odd to pass an empty string when no categories are set though. 
Have to check why `categories` can't just be an empty list when not set otherwise.

File: gradle-android-command-plugin/src/main/groovy/com/novoda/gradle/command/Monkey.groovy (hunk 10 above)
Comment (devisnik) [inline gradle-android-command-plugin/src/main/groovy/com/novoda/gradle/command/Monkey.groovy]: Yes, just iterate over the categories, but return and array of distinct (whitespace-free) parameters instead of joining them here. 
Looking at this code again, I see that you treat `categories` as a list usually, but give it the default of an empty string. Setting the default to empty list (`[]`) seems more natural and might avoid setting `-c` at all when no categories are specified.

Comment (jackSzm) [inline gradle-android-command-plugin/src/main/groovy/com/novoda/gradle/command/Monkey.groovy]: Is there any groovy closure so we can make it a one-liner like we did it with `collect {}` ?

File: gradle-android-command-plugin/src/test/groovy/com/novoda/gradle/command/AndroidCommandPluginExtensionTest.groovy (hunk 15 above)
Comment (devisnik) [inline gradle-android-command-plugin/src/test/groovy/com/novoda/gradle/command/AndroidCommandPluginExtensionTest.groovy]: I just double-checked, you can just return `null` instead of `CATEGORIES_DEFAULT`, both leads to no `-c` option being passed into the monkey command.

File: gradle-android-command-plugin/src/main/groovy/com/novoda/gradle/command/Monkey.groovy (hunk 10 above)
Comment (devisnik) [inline gradle-android-command-plugin/src/main/groovy/com/novoda/gradle/command/Monkey.groovy]:
```groovy
        getCategories().collect { ['-c', it] }.flatten()
```

Comment (jackSzm) [inline gradle-android-command-plugin/src/main/groovy/com/novoda/gradle/command/Monkey.groovy]: ok cool. I'll create another PR for all this changes.

Comment (devisnik) [inline gradle-android-command-plugin/src/main/groovy/com/novoda/gradle/command/Monkey.groovy]: To clarify about the default value (see comment above). `[]` does not work, but using `null` seems to work fine. I'd suggest to inline it into the getter then.
